pkg/operator/controller/nodeobservabilityrun/controller.go

```diff
@@ -89,92 +90,95 @@ func (r *NodeObservabilityRunReconciler) Reconcile(ctx context.Context, req ctrl
+			r.Log.Error(errUpdate, "failed to update status")
 			err = utilerrors.NewAggregate([]error{err, errUpdate})
 		}
 	}()
 
+	var isAdopted bool
+	if isAdopted, err = r.adoptResource(ctx, instance); !isAdopted {
+		if err != nil {
+			r.Log.Error(err, "Error while setting owner reference NodeObservabilityRun->NodeObservability")
+			return
+		}
+		res = ctrl.Result{Requeue: true}
+		return
+	}
+
+	var canProceed bool
+	if canProceed, err = r.preconditionsMet(ctx, instance, r.Namespace); !canProceed {
+		if err != nil {
+			r.Log.Error(err, "preconditions not met")
+			return
+		}
+		msg = fmt.Sprintf("Waiting for NodeObservabilityMachineConfig %s to become ready", instance.Spec.NodeObservabilityRef.Name)
+		instance.Status.SetCondition(nodeobservabilityv1alpha1.DebugReady, metav1.ConditionFalse, nodeobservabilityv1alpha1.ReasonInProgress, msg)
+		return ctrl.Result{RequeueAfter: pollingPeriod}, err
+	}
+	msg = "Ready to start profiling"
+	instance.Status.SetCondition(nodeobservabilityv1alpha1.DebugReady, metav1.ConditionTrue, nodeobservabilityv1alpha1.ReasonReady, msg)
+
 	if inProgress(instance) {
 		r.Log.V(3).Info("Run is in progress")
 		var requeue bool
 		requeue, err = r.handleInProgress(instance)
 		if requeue {
+			msg = "Profiling query in progress"
+			instance.Status.SetCondition(nodeobservabilityv1alpha1.DebugFinished, metav1.ConditionFalse, nodeobservabilityv1alpha1.ReasonInProgress, msg)
 			return ctrl.Result{RequeueAfter: pollingPeriod}, err
 		}
 		t := metav1.Now()
 		instance.Status.FinishedTimestamp = &t
+		msg = "Profiling query done"
+		instance.Status.SetCondition(nodeobservabilityv1alpha1.DebugFinished, metav1.ConditionTrue, nodeobservabilityv1alpha1.ReasonFinished, msg)
 		return
 	}
 
 	err = r.startRun(ctx, instance)
 	if err != nil {
+		msg = fmt.Sprintf("Failed to initiated profiling query: %s", err.Error())
+		instance.Status.SetCondition(nodeobservabilityv1alpha1.DebugFinished, metav1.ConditionFalse, nodeobservabilityv1alpha1.ReasonFailed, msg)
 		return ctrl.Result{}, err
 	}
 
+	msg = "Profiling query initiated"
+	instance.Status.SetCondition(nodeobservabilityv1alpha1.DebugFinished, metav1.ConditionFalse, nodeobservabilityv1alpha1.ReasonInProgress, msg)
+
 	// one cycle takes cca 30s
 	return ctrl.Result{RequeueAfter: time.Second * 30}, err
 }
 
+// adoptResource sets OwnerReference to point to NodeObservability from .spec.ref.name
+// returns true if already adopted, false otherwise
+func (r *NodeObservabilityRunReconciler) adoptResource(ctx context.Context, instance *nodeobservabilityv1alpha1.NodeObservabilityRun) (bool, error) {
+	for _, ref := range instance.OwnerReferences {
+		// FIXME: check kind as well?
+		if ref.Name == instance.Spec.NodeObservabilityRef.Name {
+			return true, nil
+		}
+	}
+	nodeObs := &nodeobservabilityv1alpha1.NodeObservability{}
+	if err := r.Get(ctx, types.NamespacedName{Name: instance.Spec.NodeObservabilityRef.Name}, nodeObs); err != nil {
```

Comment:
Why this API field is ever needed if the owner reference is there?

Reply:
work on removing it in #62 (not pushed yet)